lib/aPI.ml

```diff
@@ -22,9 +8,79 @@ include Rpc
 type string_list = string list [@@deriving rpc]
 
 module Ref = struct
-  type 'a t=string
-  let of_string x = x
+  (* BEGIN ref.ml from xen-api repository *)
+  type 'a t = string
+  
+  let ref_prefix = "OpaqueRef:"
+  
+  let make () =
+    let uuid = Uuidm.v `V4 |> Uuidm.to_string in
+    ref_prefix ^ uuid
+  
+  let null = ref_prefix ^ "NULL"
+  
   let string_of x = x
+  
+  let of_string x = x
+  
+  (* very ugly hack to be able to distinguish dummy and real tasks. Moreover, dummy tasks have their name *)
+  (* embedded into the reference .... *)
+  
+  (* a dummy reference is a reference of an object which is not in database *)
+  let dummy_sep = "|"
+  let dummy_prefix = "DummyRef:"
+  
+  
+  let make_dummy task_name =
+    let uuid = Uuidm.v `V4 |> Uuidm.to_string in
+    dummy_prefix ^ dummy_sep ^ uuid ^ dummy_sep ^ task_name
+  
+  let is_dummy x =
+    Astring.String.is_prefix ~affix:dummy_prefix x
+  
+  let cut_in_three ~sep str =
+    let (>>=) opt f =
+      match opt with
+      | Some value -> f value
+      | None -> None
+    in
+    let open Astring in
+    String.cut ~sep str
+    >>= fun (head, tail) ->
+      String.cut ~sep tail
```

Comment:
Very minor point, but is there a specific reason this code does not follow the usual indentation?:
```ocaml
    >>= fun (head', tail) ->
    Some (head, head', tail)
```

Reply:
Probably I forgot to ocp-indent it. I'll add a commit to fix indentation